lib/Serialization/Serialization.cpp

```diff
@@ -619,9 +619,9 @@ DeclID Serializer::addDeclRef(const Decl *D, bool allowTypeAliasXRef) {
           isa<PrecedenceGroupDecl>(D)) &&
          "cannot cross-reference this decl");
 
-  assert((!D || !isDeclXRef(D) ||
```

Comment:
Eventually this needs addressing. The assert was introduced by https://github.com/apple/swift/pull/17223, which as I understand it was designed to combat issues with ambiguous protocol conformances in SourceKit (ie. a protocol and forward declared interface share a name). 

The assert is tripped during serialization of SILModule that references the forward declared type from another module. In this case, the declaration exists, is "cross reference", and has the `ForbidSerializationReferenceAttr` attribute. Perhaps these declarations need not be included in the serialized result, but I'm not sure of the implications. If such a workaround is not possible, perhaps a diagnostic suggesting a source level fix is appropriate.